src_py/sprite.py

```diff
@@ -853,7 +895,7 @@ def change_layer(self, sprite, new_layer):
             mid += 1
         sprites.insert(mid, sprite)
         if hasattr(sprite, 'layer'):
-            sprite.layer = new_layer
+            setattr(sprite, '_layer', new_layer)
```

Comment:
Is this correct? Before it was assigning to `layer` and now to `_layer`.

Reply:
It's a little confusing because this is the only place in the original *code* that actually uses `.layer` as opposed to `._layer` so I think it was actually a bug. Made more confusing by the docs which interchangeably refer to both `.layer` and `._layer`. 

Poking through the docs again I believe it was intended that users could set the `._layer` value of a sprite **before** it is added to a group and that value would then be respected when the sprite was added, but then after entering a group the `._layer` should be read only and not settable for users. 

I suppose I could try to encode that idea into a setter for the new `.layer` property? So that it checks to see if the sprite is in any group first, and if so refuses to set the layer (as it works now) but otherwise allows it

I suspect the person/persons writing the code and docs was very confused at some point about `._layer` and `.layer` or they originally had it one way and then changed it to the other, as right now the current state is very confusing. See this from the `LayeredUpdates` class docs:

```
This group is fully compatible with pygame.sprite.Sprite class for visible game objects..

You can set the default layer through kwargs using 'default_layer' and an integer for the layer. The default layer is 0.

If the sprite you add has an attribute `_layer` then that layer will be used. If the **kwarg contains 'layer' then the sprites passed will be added to that layer (overriding the `sprite.layer` attribute). If neither sprite has attribute `layer` nor **kwarg then the default layer is used to add the sprites.
```

I mean they switch between talking about `._layer` and `.layer` from one sentence to the next . The code was actually using `._layer` not `.layer` everywhere sprites are added so setting `.layer` would do basically nothing in the current code - except that line you highlight would be activated meaning when you changed a sprite's layer (but not when you originally add it) the sprite's new `.layer` attribute would be kept up to date.

My feeling is that this code was originally written using `.layer` and then at some point in its development it was decided to change to `._layer` and several places in the docs were missed and this one place in the code. Though I could be mistaken and there is some other more confusing purpose for this. 

TLDR

- Original code is slightly confused between `._layer` (used the most times in the code and thus the most functional attribute) and `.layer` (used once in the code in a way that is mainly cosmetic).
- Original docs are confused as well using `._layer` and `.layer` seemingly interchangeably to describe behaviour that is governed solely by `._layer`.
- Making things even more confusing there are two sprite classes. One called `DirtySprite` defines `._layer` by default and the other, called `Sprite` does not, but both can be added to groups with layers in and in some cases adding a regular Sprite to a layered group will set the `._layer` property on the Sprite but not always.
- New code somewhat resolves this ambiguity by leaving `._layer` in place but adding a python property called `.layer` that returns the value of `.layer`. Right now it is read only which is consistent with the way the code works once a sprite is in a group.

**To Do**

- I think I'm going to change both of the `if hasattr(sprite, 'layer'):` lines to read `if hasattr(sprite, '_layer'):` as I think that will more accurately preserve the spirit of the original code where you could potentially have a regular Sprite object added to a layered group that didn't itself store any layer data, so long as you specified a layer when adding it to the group. This will also resolving the original bug where the code was confused here between `._layer` and `.layer`.
- I will probably also add a setter to the new `.layer` properties that sets `._layer` so long as the Sprite or DirtySprite is not already in a group. Once they are in a group they will become read only an produce an error if you try to set them.

Reply:
Yes, I see. Thanks for the description.